### ISSUE TITLE
galsim installation on Mac OSX 10.12

### ISSUE THREAD

**Issue by rmandelb** [GalSim-developers/GalSim#896]
I recently (and perhaps foolishly) went to OSX 10.12 on my Mac, and am having some GalSim installation woes.  Note that I jumped directly from 10.9 because I had been worried about 10.10 when it came out, then I just kind of settled into a state of non-updating inertia for quite a while.  So if there are new issues for GalSim installation that came up at any point in 10.10, 10.11, or 10.12, they could all be relevant here. 

I ended up re-installing TMV 0.74 (previous was 0.72), boost 1.61 (previous was 1.49), and FFTW because the new OS seems to have caused all kinds of interesting things to happen on my system.

Initially I was seeing some issue with the GalSim installation that led me to follow the advice for boost that was given here: https://github.com/GalSim-developers/GalSim/wiki/Installation%20FAQ#what-should-i-do-about-undefined-symbols-for-architecture-x86_64-errors

When compiling, it was OK with the stuff in src/, though it did give me the same warning I mentioned on the TMV issues page:
```
clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
```

Then once it went to the stuff in pysrc, I got a whole load of complaints that look like
```
ld: warning: object file (/usr/local/lib/libfftw3.a(apiplan.o)) was built for newer OSX version (10.12) than being linked (10.5)
```
Does this mean it's linking to the wrong FFTW?  I looked up at the top and it said it found FFTW, but it didn't say where.   I don't understand why 10.5 would even be relevant, since I was on OSX 10.9 before this.

But those were warnings, so it didn't die.  Moving on:

```
clang++ -o galsim/_galsim.so -bundle pysrc/.obj/module.os pysrc/.obj/Angle.os pysrc/.obj/Bounds.os pysrc/.obj/PhotonArray.os pysrc/.obj/Image.os pysrc/.obj/SBProfile.os pysrc/.obj/SBAdd.os pysrc/.obj/SBConvolve.os pysrc/.obj/SBDeconvolve.os pysrc/.obj/SBFourierSqrt.os pysrc/.obj/SBTransform.os pysrc/.obj/SBBox.os pysrc/.obj/SBGaussian.os pysrc/.obj/SBDeltaFunction.os pysrc/.obj/SBExponential.os pysrc/.obj/SBSersic.os pysrc/.obj/SBMoffat.os pysrc/.obj/SBAiry.os pysrc/.obj/SBShapelet.os pysrc/.obj/SBInterpolatedImage.os pysrc/.obj/SBKolmogorov.os pysrc/.obj/SBSpergel.os pysrc/.obj/SBInclinedExponential.os pysrc/.obj/SBInclinedSersic.os pysrc/.obj/Random.os pysrc/.obj/Noise.os pysrc/.obj/HSM.os pysrc/.obj/Integ.os pysrc/.obj/Table.os pysrc/.obj/Interpolant.os pysrc/.obj/CorrelatedNoise.os pysrc/.obj/Bessel.os pysrc/.obj/CDModel.os pysrc/.obj/Silicon.os pysrc/.obj/RealGalaxy.os pysrc/.obj/WCS.os -L/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/config -Llib -L/usr/local/lib -lboost_python -lpython2.7 -ltmv_symband -lfftw3 -ltmv -lblas -lgalsim
clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
Undefined symbols for architecture x86_64:
  "boost::python::objects::function_object(boost::python::objects::py_function const&, std::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&)", referenced from:
      void boost::python::detail::def_init_aux<boost::python::class_<galsim::AngleUnit, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>, boost::python::default_call_policies, boost::mpl::vector1<double>, boost::mpl::size<boost::mpl::vector1<double> > >(boost::python::class_<galsim::AngleUnit, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>&, boost::mpl::vector1<double> const&, boost::mpl::size<boost::mpl::vector1<double> >, boost::python::default_call_policies const&, char const*, std::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&) in Angle.os
```
and more like that, followed by
```
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [galsim/_galsim.so] Error 129
```

I did follow the instructions already on how to recompile boost.  (And I did the usual of explicitly telling it where my python lives, as well.)  Do you think this is definitely about boost or could it be related to the FFTW warning or ...? Thoughts?

**Comment by rmjarvis** [comment]
I'm on 10.10 myself, so might not have a huge amount of helpful advice here.  But that said, I'll try.

>  I looked up at the top and it said it found FFTW, but it didn't say where. I don't understand why 10.5 would even be relevant, since I was on OSX 10.9 before this.

I think this is probably important.  You likely need to reinstall fftw.  As for why 10.5 rather than 10.9, my guess is that the ABI compatibility didn't change between 10.5 and 10.9, so they probably just called it 10.5 for all in between.  Then sometime between 10.9 and 10.12, they probably changed something that made it non-ABI-compatible.  In any case, probably best to delete any libfftw.a files that are on your system from the old OS and reinstall new ones.

> clang: error: linker command failed with exit code 1 (use -v to see invocation)

Sounds like something with the Boost linking.  Did you remove the old version of boos from your system?  There's no point keeping it around, so if you didn't, probably a good idea.  Otherwise it's possible that the linker isn't finding the right boost library to link to.

You can also have it report the libraries being linked by running the linking command yourself with verbose output.
```
clang++ -v -Wl,--verbose -o galsim/_galsim.so -bundle ...
```
Then you can check that it's using the right versions of boost, fftw, and tmv.

**Comment by rmandelb** [comment]
> I think this is probably important. You likely need to reinstall fftw.

But I did already?  (TMV, boost, and fftw were the things I reinstalled before even attempting GalSim)  If I look in /usr/local/*, where I keep all the software that I install myself, all the fftw stuff is new. 

> Sounds like something with the Boost linking. Did you remove the old version of boos from your 
> system?

I did not, but scons said
```
Boost version is 1.61.0
```
(that's the new one).   Likewise the libraries and include files under /usr/local appear to be new.  I did try to see if there was some boost command ```./b2 BLAH``` that could be used to remove old boost files but there didn't seem to be.  I could manually go into /usr/local and ```sudo rm``` the boost files that are old, unless you know of some nice automatic way to do it?

Regarding your last suggestion, clang++ is insisting it doesn't have a --verbose option:
```
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -dynamic -arch x86_64 -bundle -macosx_version_min 10.12.0 -syslibroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk -o galsim/_galsim.so -L/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/config -Llib -L/usr/local/lib --verbose pysrc/.obj/module.os pysrc/.obj/Angle.os pysrc/.obj/Bounds.os pysrc/.obj/PhotonArray.os pysrc/.obj/Image.os pysrc/.obj/SBProfile.os pysrc/.obj/SBAdd.os pysrc/.obj/SBConvolve.os pysrc/.obj/SBDeconvolve.os pysrc/.obj/SBFourierSqrt.os pysrc/.obj/SBTransform.os pysrc/.obj/SBBox.os pysrc/.obj/SBGaussian.os pysrc/.obj/SBDeltaFunction.os pysrc/.obj/SBExponential.os pysrc/.obj/SBSersic.os pysrc/.obj/SBMoffat.os pysrc/.obj/SBAiry.os pysrc/.obj/SBShapelet.os pysrc/.obj/SBInterpolatedImage.os pysrc/.obj/SBKolmogorov.os pysrc/.obj/SBSpergel.os pysrc/.obj/SBInclinedExponential.os pysrc/.obj/SBInclinedSersic.os pysrc/.obj/Random.os pysrc/.obj/Noise.os pysrc/.obj/HSM.os pysrc/.obj/Integ.os pysrc/.obj/Table.os pysrc/.obj/Interpolant.os pysrc/.obj/CorrelatedNoise.os pysrc/.obj/Bessel.os pysrc/.obj/CDModel.os pysrc/.obj/Silicon.os pysrc/.obj/RealGalaxy.os pysrc/.obj/WCS.os -lboost_python -lpython2.7 -ltmv_symband -lfftw3 -ltmv -lblas -lgalsim -lc++ -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.1.0/lib/darwin/libclang_rt.osx.a
ld: unknown option: --verbose
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by rmjarvis** [comment]
> I could manually go into /usr/local and sudo rm the boost files that are old, unless you know of some nice automatic way to do it?

That's what I usually do if I need to remove the old ones.

> Regarding your last suggestion, ld is insisting it doesn't have a --verbose option:

Sorry, I guess the Apple `ld` command just calls it `-v`, so the command would be
```
clang++ -v -Wl,-v ...
```

**Comment by rmandelb** [comment]
> That's what I usually do if I need to remove the old ones.

Yuck.  Since there were some old boost files in /usr/local (any that didn't get updated when I installed boost python and math for 1.61; mostly non-python and non-math ones were left from what I could see) I removed them all and reinstalled boost math and python cleanly.

Thanks for the fix on the verbose command.  I'm still not finding much useful in the output, even though there's a lot more of it.  For example, it starts out like this:
```
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -dynamic -arch x86_64 -bundle -macosx_version_min 10.12.0 -syslibroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk -o galsim/_galsim.so -L/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/config -Llib -L/usr/local/lib -v pysrc/.obj/module.os pysrc/.obj/Angle.os pysrc/.obj/Bounds.os pysrc/.obj/PhotonArray.os pysrc/.obj/Image.os pysrc/.obj/SBProfile.os pysrc/.obj/SBAdd.os pysrc/.obj/SBConvolve.os pysrc/.obj/SBDeconvolve.os pysrc/.obj/SBFourierSqrt.os pysrc/.obj/SBTransform.os pysrc/.obj/SBBox.os pysrc/.obj/SBGaussian.os pysrc/.obj/SBDeltaFunction.os pysrc/.obj/SBExponential.os pysrc/.obj/SBSersic.os pysrc/.obj/SBMoffat.os pysrc/.obj/SBAiry.os pysrc/.obj/SBShapelet.os pysrc/.obj/SBInterpolatedImage.os pysrc/.obj/SBKolmogorov.os pysrc/.obj/SBSpergel.os pysrc/.obj/SBInclinedExponential.os pysrc/.obj/SBInclinedSersic.os pysrc/.obj/Random.os pysrc/.obj/Noise.os pysrc/.obj/HSM.os pysrc/.obj/Integ.os pysrc/.obj/Table.os pysrc/.obj/Interpolant.os pysrc/.obj/CorrelatedNoise.os pysrc/.obj/Bessel.os pysrc/.obj/CDModel.os pysrc/.obj/Silicon.os pysrc/.obj/RealGalaxy.os pysrc/.obj/WCS.os -lboost_python -lpython2.7 -ltmv_symband -lfftw3 -ltmv -lblas -lgalsim -lc++ -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.1.0/lib/darwin/libclang_rt.osx.a
@(#)PROGRAM:ld  PROJECT:ld64-278.4
configured to support archs: armv6 armv7 armv7s arm64 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em (tvOS)
Library search paths:
	/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/config
	lib
	/usr/local/lib
	/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/lib
```

The only one of these library search paths that I install stuff into is /usr/local/lib.  If I do ```ls -la /usr/local/lib/*fftw*``` the output is
```
-rw-r--r--+ 1 root  admin  1531896 Jul 30 14:08 /usr/local/lib/libfftw3.a
-rwxr-xr-x+ 1 root  admin      892 Jul 30 14:08 /usr/local/lib/libfftw3.la
```
Those are the new ones I installed earlier this afternoon.

If i do the same for boost, I get
```
-rwxr-xr-x+ 1 root  admin  110248 Jul 30 17:07 /usr/local/lib/libboost_math_c99.dylib
-rwxr-xr-x+ 1 root  admin   72336 Jul 30 17:07 /usr/local/lib/libboost_math_c99f.dylib
-rwxr-xr-x+ 1 root  admin   84972 Jul 30 17:07 /usr/local/lib/libboost_math_c99l.dylib
-rwxr-xr-x+ 1 root  admin  469164 Jul 30 17:05 /usr/local/lib/libboost_math_tr1.dylib
-rwxr-xr-x+ 1 root  admin  467328 Jul 30 17:06 /usr/local/lib/libboost_math_tr1f.dylib
-rwxr-xr-x+ 1 root  admin  452320 Jul 30 17:06 /usr/local/lib/libboost_math_tr1l.dylib
-rwxr-xr-x+ 1 root  admin  320484 Jul 30 17:08 /usr/local/lib/libboost_python.dylib
```
Time stamps are all from a few minutes ago because I just did a re-install after removing all old files from the previous boost version.

So there is nothing obvious in where it's looking for libraries that should be a problem.  I put the file at https://www.dropbox.com/s/yf20padxh5h3vry/clang_ld_verbose_output.txt?dl=0 in case you can see anything obvious?

**Comment by rmjarvis** [comment]
My best guess is that Boost was compiled with c++11 support, but we use -std=c++98.  A lot of those errors look like std library stuff and I know sometimes they are incompatible between c++11/14/17 and c++98.  

Try checking if Boost has an option to use c++98 instead.  (And recompile it with that.)

**Comment by rmandelb** [comment]
Thanks - that gives me something to go on!  No time to play with this for the next day or so, but that's the next thing I will try.

**Comment by josePhoenix** [comment]
Well, good to see I'm not the only one. I have more or less the same issue.

**Comment by josePhoenix** [comment]
Actually, I don't have any messages that reference fftw. Here's my build output:

```
$ scons PREFIX=$PYHOME PYTHON=$PYHOME/bin/python PYPREFIX=$PYHOME/lib/python3.6/site-packages BOOST_DIR=$PYHOME
scons: Reading SConscript files ...
SCons is version 2.5.1 using python version 2.7.13
Python is from /Users/jlong/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/include/python2.7
Using the following (non-default) scons options:
   PYTHON = /Users/jlong/miniconda2/envs/galsim-master/bin/python
   PREFIX = /Users/jlong/miniconda2/envs/galsim-master
   PYPREFIX = /Users/jlong/miniconda2/envs/galsim-master/lib/python3.6/site-packages
   BOOST_DIR = /Users/jlong/miniconda2/envs/galsim-master
   EXTRA_INCLUDE_PATH = /Users/jlong/include
   EXTRA_LIB_PATH = /Users/jlong/lib
   EXTRA_PATH = /Users/jlong/bin
These can be edited directly in the file gs_scons.conf.
Type scons -h for a full list of available options.
Using python =  /Users/jlong/miniconda2/envs/galsim-master/bin/python
Using compiler: /usr/bin/g++
compiler version: 4.2.1
Determined that a good number of jobs = 8
Using SSE with flag -msse2
Checking for C++ header file fftw3.h... (cached) yes
Checking for correct FFTW linkage... (cached) yes
Checking for boost header files... (cached) yes
Boost version is 1.61.0
Checking for C++ header file TMV.h... yes
TMV version is 0.74
Using TMV_LINK file: /usr/local/share/tmv/tmv-link
     -L/usr/local/lib -ltmv
Mac version is 10.11.6
XCode version is 8.1
Checking for correct TMV linkage... (this may take a little while)
Checking for correct TMV linkage... yes
Checking if we can build against Python... yes
Building for python version 3.6
Checking if we can build module using TMV... yes
Checking if we can build against NumPy... yes
Numpy version is 1.13.1
Checking for PyFITS... (cached) yes
Astropy version is 2.0
Checking for future... (cached) yes
Future version is 0.16.0
Checking if we can build against Boost.Python... yes
Checking if C++ exceptions are propagated up to python... yes
GalSim version  1.5
scons: done reading SConscript files.
scons: Building targets ...
g++ -o src/.obj/FFT.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/FFT.cpp
g++ -o src/.obj/Laguerre.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/Laguerre.cpp
g++ -o src/.obj/SBProfile.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBProfile.cpp
g++ -o src/.obj/SBBox.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBBox.cpp
g++ -o src/.obj/SBGaussian.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBGaussian.cpp
g++ -o src/.obj/SBDeltaFunction.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBDeltaFunction.cpp
g++ -o src/.obj/SBExponential.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBExponential.cpp
g++ -o src/.obj/SBSersic.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBSersic.cpp
g++ -o src/.obj/SBMoffat.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBMoffat.cpp
g++ -o src/.obj/SBAiry.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBAiry.cpp
g++ -o src/.obj/SBShapelet.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBShapelet.cpp
g++ -o src/.obj/SBAdd.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBAdd.cpp
g++ -o src/.obj/SBConvolve.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBConvolve.cpp
g++ -o src/.obj/SBTransform.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBTransform.cpp
g++ -o src/.obj/SBInterpolatedImage.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBInterpolatedImage.cpp
g++ -o src/.obj/SBDeconvolve.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBDeconvolve.cpp
g++ -o src/.obj/SBFourierSqrt.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBFourierSqrt.cpp
g++ -o src/.obj/SBKolmogorov.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBKolmogorov.cpp
g++ -o src/.obj/SBSpergel.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBSpergel.cpp
g++ -o src/.obj/SBInclinedExponential.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBInclinedExponential.cpp
g++ -o src/.obj/SBInclinedSersic.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/SBInclinedSersic.cpp
g++ -o src/.obj/Table.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/Table.cpp
g++ -o src/.obj/CorrelatedNoise.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/CorrelatedNoise.cpp
g++ -o src/.obj/RealGalaxy.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/RealGalaxy.cpp
g++ -o src/.obj/WCS.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/WCS.cpp
g++ -o src/hsm/.obj/PSFCorr.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include src/hsm/PSFCorr.cpp
g++ -o bin/sizeof_SIFD src/.obj/sizeof_SIFD.o -Llib -L/Users/jlong/miniconda2/envs/galsim-master/lib -L/Users/jlong/lib -L/usr/local/lib -ltmv_symband -lfftw3 -ltmv
g++ -o pysrc/.obj/SBShapelet.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include -I/Users/jlong/miniconda2/envs/galsim-master/include/python3.6m -I/Users/jlong/miniconda2/envs/galsim-master/lib/python3.6/site-packages/numpy/core/include pysrc/SBShapelet.cpp
g++ -o pysrc/.obj/SBInterpolatedImage.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include -I/Users/jlong/miniconda2/envs/galsim-master/include/python3.6m -I/Users/jlong/miniconda2/envs/galsim-master/lib/python3.6/site-packages/numpy/core/include pysrc/SBInterpolatedImage.cpp
g++ -o pysrc/.obj/CorrelatedNoise.os -c -O2 -std=c++98 -fno-strict-aliasing -msse2 -fPIC -Iinclude/galsim -Iinclude -I/Users/jlong/miniconda2/envs/galsim-master/include -I/Users/jlong/include -I/Users/jlong/miniconda2/envs/galsim-master/include/python3.6m -I/Users/jlong/miniconda2/envs/galsim-master/lib/python3.6/site-packages/numpy/core/include pysrc/CorrelatedNoise.cpp
BuildExecutableScript(["bin/galsim"], ["bin/galsim.py"])
BuildExecutableScript(["bin/galsim_yaml"], ["bin/galsim_yaml.py"])
BuildExecutableScript(["bin/galsim_json"], ["bin/galsim_json.py"])
BuildExecutableScript(["bin/galsim_download_cosmos"], ["bin/galsim_download_cosmos.py"])
BuildExecutableScript(["bin/installed_python"], ["bin/installed_python.py"])
g++ -o lib/libgalsim.1.5.dylib -dynamiclib -dynamic -Wl,-install_name,/Users/jlong/miniconda2/envs/galsim-master/lib/libgalsim.1.5.dylib -Wl,-compatibility_version,1.5 -Wl,-current_version,1.5.0 src/.obj/BinomFact.os src/.obj/FFT.os src/.obj/Image.os src/.obj/Interpolant.os src/.obj/Laguerre.os src/.obj/OneDimensionalDeviate.os src/.obj/PhotonArray.os src/.obj/GSParams.os src/.obj/SBProfile.os src/.obj/SBBox.os src/.obj/SBGaussian.os src/.obj/SBDeltaFunction.os src/.obj/SBExponential.os src/.obj/SBSersic.os src/.obj/SBMoffat.os src/.obj/SBAiry.os src/.obj/SBShapelet.os src/.obj/SBAdd.os src/.obj/SBConvolve.os src/.obj/SBTransform.os src/.obj/SBInterpolatedImage.os src/.obj/SBDeconvolve.os src/.obj/SBFourierSqrt.os src/.obj/SBKolmogorov.os src/.obj/SBSpergel.os src/.obj/SBInclinedExponential.os src/.obj/SBInclinedSersic.os src/.obj/Table.os src/.obj/RealSpaceConvolve.os src/.obj/Random.os src/.obj/CorrelatedNoise.os src/.obj/CDModel.os src/.obj/Version.os src/.obj/Polygon.os src/.obj/Silicon.os src/.obj/RealGalaxy.os src/.obj/WCS.os src/hsm/.obj/PSFCorr.os -Llib -L/Users/jlong/miniconda2/envs/galsim-master/lib -L/Users/jlong/lib -L/usr/local/lib -ltmv_symband -lfftw3 -ltmv
SymLink(["lib/libgalsim.dylib"], ["lib/libgalsim.1.5.dylib"])
g++ -o galsim/_galsim.so -bundle pysrc/.obj/module.os pysrc/.obj/Angle.os pysrc/.obj/Bounds.os pysrc/.obj/PhotonArray.os pysrc/.obj/Image.os pysrc/.obj/SBProfile.os pysrc/.obj/SBAdd.os pysrc/.obj/SBConvolve.os pysrc/.obj/SBDeconvolve.os pysrc/.obj/SBFourierSqrt.os pysrc/.obj/SBTransform.os pysrc/.obj/SBBox.os pysrc/.obj/SBGaussian.os pysrc/.obj/SBDeltaFunction.os pysrc/.obj/SBExponential.os pysrc/.obj/SBSersic.os pysrc/.obj/SBMoffat.os pysrc/.obj/SBAiry.os pysrc/.obj/SBShapelet.os pysrc/.obj/SBInterpolatedImage.os pysrc/.obj/SBKolmogorov.os pysrc/.obj/SBSpergel.os pysrc/.obj/SBInclinedExponential.os pysrc/.obj/SBInclinedSersic.os pysrc/.obj/Random.os pysrc/.obj/Noise.os pysrc/.obj/HSM.os pysrc/.obj/Integ.os pysrc/.obj/Table.os pysrc/.obj/Interpolant.os pysrc/.obj/CorrelatedNoise.os pysrc/.obj/Bessel.os pysrc/.obj/CDModel.os pysrc/.obj/Silicon.os pysrc/.obj/RealGalaxy.os pysrc/.obj/WCS.os -Llib -L/Users/jlong/miniconda2/envs/galsim-master/lib -L/Users/jlong/lib -L/usr/local/lib -lboost_python3 -lpython3.6m -ltmv_symband -lfftw3 -ltmv -lgalsim
Undefined symbols for architecture x86_64:
  "boost::python::objects::function_object(boost::python::objects::py_function const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&)", referenced from:
      boost::python::api::object boost::python::detail::make_function_aux<void (*)(_object*, double), boost::python::default_call_policies, boost::mpl::vector3<void, _object*, double>, mpl_::int_<0> >(void (*)(_object*, double), boost::python::default_call_policies const&, boost::mpl::vector3<void, _object*, double> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<_object* (*)(galsim::AngleUnit&, galsim::AngleUnit const&), boost::python::default_call_policies, boost::mpl::vector3<_object*, galsim::AngleUnit&, galsim::AngleUnit const&>, mpl_::int_<0> >(_object* (*)(galsim::AngleUnit&, galsim::AngleUnit const&), boost::python::default_call_policies const&, boost::mpl::vector3<_object*, galsim::AngleUnit&, galsim::AngleUnit const&> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<double (galsim::AngleUnit::*)() const, boost::python::default_call_policies, boost::mpl::vector2<double, galsim::AngleUnit&>, mpl_::int_<0> >(double (galsim::AngleUnit::*)() const, boost::python::default_call_policies const&, boost::mpl::vector2<double, galsim::AngleUnit&> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<_object* (*)(galsim::AngleUnit&, double const&), boost::python::default_call_policies, boost::mpl::vector3<_object*, galsim::AngleUnit&, double const&>, mpl_::int_<0> >(_object* (*)(galsim::AngleUnit&, double const&), boost::python::default_call_policies const&, boost::mpl::vector3<_object*, galsim::AngleUnit&, double const&> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<double (*)(galsim::AngleUnit, galsim::AngleUnit), boost::python::default_call_policies, boost::mpl::vector3<double, galsim::AngleUnit, galsim::AngleUnit>, mpl_::int_<0> >(double (*)(galsim::AngleUnit, galsim::AngleUnit), boost::python::default_call_policies const&, boost::mpl::vector3<double, galsim::AngleUnit, galsim::AngleUnit> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<void (*)(_object*), boost::python::default_call_policies, boost::mpl::vector2<void, _object*>, mpl_::int_<0> >(void (*)(_object*), boost::python::default_call_policies const&, boost::mpl::vector2<void, _object*> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<void (*)(_object*, double, galsim::AngleUnit), boost::python::default_call_policies, boost::mpl::vector4<void, _object*, double, galsim::AngleUnit>, mpl_::int_<0> >(void (*)(_object*, double, galsim::AngleUnit), boost::python::default_call_policies const&, boost::mpl::vector4<void, _object*, double, galsim::AngleUnit> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      ...
  "boost::python::objects::register_dynamic_id_aux(boost::python::type_info, std::__1::pair<void*, boost::python::type_info> (*)(void*))", referenced from:
      galsim::pyExportAngle() in Angle.os
      boost::python::class_<galsim::AngleUnit, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, boost::python::no_init_t) in Angle.os
      galsim::pyExportBounds() in Bounds.os
      boost::python::class_<galsim::Position<double>, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, boost::python::no_init_t) in Bounds.os
      boost::python::class_<galsim::Position<int>, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, boost::python::no_init_t) in Bounds.os
      boost::python::class_<galsim::PhotonArray, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, boost::python::no_init_t) in PhotonArray.os
      boost::python::class_<galsim::BaseImage<unsigned short>, boost::noncopyable_::noncopyable, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, char const*, boost::python::no_init_t) in Image.os
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [galsim/_galsim.so] Error 1
scons: building terminated because of errors.
```

**Comment by rmandelb** [comment]
I found [this thread](http://boost.2283326.n4.nabble.com/mixing-c-98-with-c-11-td4656026.html) that suggests @rmjarvis might be right that the c++11 vs. c++98 thing might be a problem (nothing particularly deep here, just "yep, that is probably going to give linking errors").

So I tried adding ```-std=c++98``` when running b2 to compile and install boost (and I used ```b2 -a``` to force it to really actually rebuild everything).  Note that I couldn't find a definitive statement that ```-std``` is actually an allowed option or that it would do anything when compiling the library (for all I know, it completely ignored me), but at least it didn't complain.   I am skeptical that it actually used the flag, because the output didn't show the flag propagating through anywhere when compiling. Then when I went back to install GalSim, I got the same exact warnings and failures as before.

For the sake of completeness, my full command that I was using is
```
./b2 -a --with-python --with-math toolset=clang cxxflags="-stdlib=libc++" linkflags="-stdlib=libc++" -std=c++98 link=shared install
```
The ```toolset```, ```cxxflags```, and ```linkflags``` came from our FAQ to address a different problem.  Do you think these flags might be incompatible with the ```-stdlib``` option?  For what it's worth, when boost is compiling, I see a lot of ```clang-darwin.compile.c++``` which is not something I've seen elsewhere.

So then I tried again (second try today) with
```
./b2 -a --with-python --with-math -std=c++98 link=shared install
```
and now I see a lot of ```darwin.compile.c++``` due to removal of the ```toolset=clang```, but the compilation commands still do not appear to be doing anything explicitly involving ```-std=c++98```.  When trying to compile GalSim with this version of boost, I get the same errors as before.

So I think the ```-std``` flag is simply being ignored.  Then I found [this thread](https://stackoverflow.com/questions/2887707/how-to-build-boost-with-c0x-support) that suggests I should be doing something like ```cxxflags="-std=c++98"```.  So I tried compiling yet again that way (it also suggested I should be including the ```toolset=clang``` option so I put that back too for this third try).  But still, I do not see the ```-std=c++98``` on any of the lines where it's compiling stuff, unlike in GalSim where it's clearly invoking that option.  And again, GalSim cannot compile, same warnings and errors as before.

Thoughts?  (Other than "boost is the worst".)  Should I be messing around with making ```user-config.jam``` files?  The stackoverflow thread seemed to suggest that for later versions of boost, like this one, the ```cxxflags``` option should work, but maybe that's just wrong?

**Comment by rmandelb** [comment]
Update: For compiling boost, I tried making a `user-config.jam` file that looks like
```
using clang
   : std98
   : "/usr/bin/clang++"
   : <cxxflags>-std=c++98
```
following what was suggested on one of the pages I linked.  When I invoked b2 with `toolset=clang-std98`, I didn't see it explicitly including the cxx flags, but there were lots of lines like
```
clang-darwin.compile.c++ bin.v2/libs/math/build/clang-darwin-std98/release/threading-multi/assoc_laguerre.o
```
which at least have something indicating std98 in them, so I thought this might be progress.  (With what I was trying before, the bit that said clang-darwin-std98 just said clang-darwin-4.2.1.)  But then I got all the same warnings and errors when compiling GalSim.  I think I need to do something else to force it to compile with `-std=c++98` but so far no variation that I've tried in `user-config.jam` worked.

**Comment by rmandelb** [comment]
I tried turning off the `-std=c++98` flag in GalSim as you suggested, but it had no effect on the warnings and errors.

I am still pretty confused about some basic stuff, even before we get to boost.  For example, the FFTW warnings:
```
ld: warning: object file (/usr/local/lib/libfftw3.a(apiplan.o)) was built for newer OSX version (10.12) than being linked (10.5)
```
Do you understand what that means?  I initially thought the linker was finding another FFTW, but I cannot find fftw library on my system besides the new one that I rebuilt a few weeks ago.  Am I misinterpreting this warning entirely?

I tried running the verbose linking command as you suggested, but I'm not finding the output terribly illuminating.  Do you have suggestions for what to look for in the output?  I dumped it [here](https://www.dropbox.com/s/h378yxusqoyy2mu/tmp.txt?dl=0) in case you want to see if anything catches your eye.

At this point I am considering doing the whole thing from scratch, by which I mean new python and everything else (I'm using a really old EPD python), but I would feel happier about doing that if I had a sense for where the problem is really arising.

**Comment by rmjarvis** [comment]
Not sure about the warning.  From Googling around a bit, it seems like people are implying that it is harmless, but it could certainly be indicating something.  I just don't know what.

Thanks for the linker output.  Not terribly helpful for me either.  It certainly looks like some kind of compiler mismatch (or -std flag mismatch).  But I don't know where the mismatch could be happening.  Looks like it has to be boost, but I don't have any other ideas about how to alter its compilation.

I think getting a fresh EPD Canopy installation would probably be a good idea.  Could be that there is something in there that is now incompatible with OSX 10.12.

**Comment by rmjarvis** [comment]
Also, maybe this snafu will encourage you to code review the open PRs for #809.  :)  Which in turn will encourage me to take the next steps to ridding GalSim of boost...

**Comment by rmandelb** [comment]
Yeah, I've decided to bite the bullet and go for a new EPD Canopy installation.  That'll be my fun project for the day.

Thanks for the subtle hint about #809.  ;)  I'll aim for this week.

**Comment by rmandelb** [comment]
OK, I'm trying to use EPD Canopy (python 2.7 because I really don't want to go to 3 just yet).

I can use the EPD Canopy and import numpy, scipy, future, matplotlib, etc.

Unfortunately GalSim refuses to use that python even though I'm specifying it and it's in my path:
```
scons: Reading SConscript files ...
SCons is version 2.1.0 using python version 2.7.10
Python is from /System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7
Using the following (non-default) scons options:
   CXX = clang++
   PYTHON = ~/Library/Enthought/Canopy/edm/envs/User/bin/python2.7
   FFTW_DIR = /usr/local
   BOOST_DIR = /usr/local/include
   EXTRA_INCLUDE_PATH = /usr/local/include
   DYLD_LIBRARY_PATH = /Users/rmandelb/Library/Enthought/Canopy/edm/envs/User/lib
   LD_LIBRARY_PATH = /Users/rmandelb/Library/Enthought/Canopy/edm/envs/User/lib:/usr/local/lib
These can be edited directly in the file gs_scons.conf.
Type scons -h for a full list of available options.

Specified python not found in path: ~/Library/Enthought/Canopy/edm/envs/User/bin/python2.7
Mac version is 10.12.6

Starting with El Capitan (OSX 10.11), Apple instituted a new policy called
"System Integrity Protection" (SIP) where they strip "dangerous" environment
variables from system calls (including SCons).  So if your system is using
DYLD_LIBRARY_PATH for run-time library resolution, then SCons cannot see it
so that may be why this is failing.  cf. Issues #721 and #725.
You should try executing:

    scons DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH

Please fix the above error(s) and rerun scons.
```
Except my DYLD_LIBRARY_PATH is not actually set, so following that advice won't help.

**Comment by rmjarvis** [comment]
It's possible the `~` is messing it up.  Did you try giving the actual path without that shorthand?  (Usually SCons expands it, but every so often I find cases where it doesn't work as I expect.  I never really figured out why it sometimes doesn't expand it.)

Also, if you don't have `DYLD_LIBRARY_PATH` set in your environment, then you probably shouldn't need to set it for GalSim either. 

I also wouldn't have thought you would need to explicitly add `/usr/bin/include` as an `EXTRA_INCLUDE_PATH`, but I suppose you probably did that for good reason, so maybe that's necessary for some reason.

**Comment by rmandelb** [comment]
Ahhhhhh it was the `~`!  Thank you!  Now it's fine with the python and fails to link TMV even though I started handing it `TMV_DIR`.  I went into config.log and noticed the following:

1. Lots of badness associate with the earlier tests of FFTW, e.g.,
```
clang++ -o .sconf_temp/conftest_4.o -c -O2 -std=c++98 -msse2 -Iinclude -I/usr/local/include .sconf_temp/conftest_4.cpp
clang++ -o .sconf_temp/conftest_4 .sconf_temp/conftest_4.o -Llib -L/usr/local/lib
Undefined symbols for architecture x86_64:
  "_fftw_destroy_plan", referenced from:
      _main in conftest_4.o
  "_fftw_free", referenced from:
      _main in conftest_4.o
  "_fftw_malloc", referenced from:
      _main in conftest_4.o
  "_fftw_plan_dft_r2c_2d", referenced from:
      _main in conftest_4.o
ld: symbol(s) not found for architecture x86_64
```

2. It finds TMV.h, but when trying to link there's a lot of stuff like this:
```
clang++ -o .sconf_temp/conftest_14.o -c -O2 -std=c++98 -msse2 -Iinclude -I/usr/local/include .sconf_temp/conftest_14.cpp
clang++ -o .sconf_temp/conftest_14 -fopenmp .sconf_temp/conftest_14.o -Llib -L/usr/local/lib -lfftw3 -ltmv -lblas -lpthread
clang: error: unsupported option '-fopenmp'
```

**Comment by rmjarvis** [comment]
> "_fftw_destroy_plan", referenced from:

As long as it eventually gets something that works, this kind of error in the early tests of FFTW is expected.  We start by trying to compile without adding the `-lfftw` link flag.  This way if the user adds something explicit to do the linking, then we want to give it a chance to work.  That normally fails.  Then it tries the different ways we know about to add the appropriate fftw library.  So as long as one of these later ones works, then it's fine.

> `unsupported option '-fopenmp'`

Did `-fopenmp` show up in the reported tmv_link flags?  It shouldn't have for clang++ unless MacOS is now shipping with the versions of clang that support it, which I guess is possible.  But in that case, it wouldn't be failing with this message, so I suspect not.  

Did you recompile TMV in all of this process?  Maybe use `CXX=clang++` there too to make sure it's not getting confused by Apple's clang masquerading as gcc.

**Comment by rmandelb** [comment]
Yes, I did recompile TMV along with many other GalSim dependencies at the start of this process, and I did not use `CXX=clang++` at that time.  I just did it again while specifying clang++. Now it's reporting success on linking to and building modules against TMV, and there are no warnings or errors in the process of compiling GalSim.  

Now that I've updated basically everything on my system, I've also gotten rid of that warning about C++-layer exceptions not getting propagated appropriately up to the python layer, which was an issue on my Mac for ages.  

When doing 'scons tests' I got
```
WARNING: Could not find pytest in path.
         The cpp test programs will be compiled and run,
         but you will have to run the python tests manually.
```
I am a little surprised; when sitting in a terminal, I can run pytest, and if I open a python interpreter I can import the pytest module.   Is this some other issue with paths not getting properly passed around or ...?

**Comment by rmjarvis** [comment]
> Could not find pytest in path.

Not sure why that might be.  You can try specifying the explicit path for pytest as `scons PYTEST=/path/to/pytest`.  But if it's in your path, then I don't have a guess about why it's not finding it automatically.

**Comment by rmandelb** [comment]
It is in my path, but I just tried passing it in directly, and that works... except now I'm getting
```
Using pytest from:  /Users/rmandelb/Library/Enthought/Canopy/edm/envs/User/bin/pytest
pytest is version 3.2.1... running tests in parallel with 8 jobs
If you have trouble with the parallel test run, you may disable it by typing:
    scons tests -j1
Also, if you get an error about unrecognized arguments, you might need to run:
    pip install pytest-xdist pytest-timeout --user

Starting python tests...
usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: -n=8 --timeout=60
  inifile: None
  rootdir: /Users/rmandelb/git/GalSim/tests
pytest returned error code  2
```

If I follow the instructions and install pytest through pip, then is that going to be a problem because it's not part of the Canopy distribution and the python I'm using is?

**Comment by rmjarvis** [comment]
There should be a pip in the Canopy distro that will install things into that, rather than the system python.  Do `which -a pip` to see all the possible pips.  

And maybe make your path such that the Canopy one is found first.  Or just make an alias so pip calls the Canopy one.

**Comment by rmjarvis** [comment]
BTW, the workaround for now is that `scons tests -j1` will avoid those pytest plugins, so that should already work.

**Comment by rmandelb** [comment]
Huh, it seems something is broken with my path, because `which -a pip` gives me nothing.  If I do `type pip` then I get `pip is hashed (/Users/rmandelb/Library/Enthought/Canopy/edm/envs/User/bin/pip)`.  And thanks for the reminder about `-j1`.

**Comment by rmjarvis** [comment]
Well, I think that means pip is the one for Canopy, so you're probably fine running pip at least.

**Comment by rmandelb** [comment]
Yep, good point.  I tried running with `--user` as suggested, but when I did that, it apparently put pytest somewhere not in my path.  I had to use `sudo` to get it to update its own pytest.

Anyway, tests pass and it's all working, so that's great.  And I got a shiny new python out of the deal.

Before I close this, I wanted to ask @josePhoenix if any of this might be helpful to him, since he posted in this thread that he was having the same problem as me?

**Comment by rmandelb** [comment]
Thanks for your help @rmjarvis !

**Comment by jmeyers314** [comment]
> BTW, the workaround for now is that scons tests -j1 will avoid those pytest plugins, so that should already work.

Ack!  I should have thought of that when switching nose -> pytest.  You can get parallel testing and timeouts back with some variant of:

`pip install pytest-xdist pytest-timeout`

I'll make a quick PR to add instructions to the README.md.

**Comment by rmjarvis** [comment]
I added it to the SCons output already.  It warns that
```
Also, if you get an error about unrecognized arguments, you might need to run:
    pip install pytest-xdist pytest-timeout --user
```
For Rachel, it was just the `--user` that failed with EPD for some reason.

**Comment by rmjarvis** [comment]
> I'll make a quick PR to add instructions to the README.md.

Not to imply that you shouldn't added it to the README though.  That's probably a good idea too.  And IMO, you don't need a PR for that kind of thing.  You can just do that on master directly if you want.

**Comment by jmeyers314** [comment]
>Not to imply that you shouldn't added it to the README though. That's probably a good idea too. And IMO, you don't need a PR for that kind of thing. You can just do that on master directly if you want.

Sure.  Will do.  Thanks for adding the SCons error message.

**Comment by rmjarvis** [comment]
Is there any action to be taken on this issue? Or can we close this?

**Comment by rmandelb** [comment]
No action to be taken.  I had left it open to see if the other person with the problem would comment as to whether it helped him or not, but it's been a few months, so will close.

**Comment by rmandelb** [comment]
Closing #896.
